### DISCUSSION TITLE
Pattern matching with UTF-8 String Literals

### DISCUSSION THREAD

**Discussion by tmds** [dotnet/csharplang#6036] [Language Ideas] [answered]
[C# 11 preview blog post](https://devblogs.microsoft.com/dotnet/csharp-11-preview-updates/) includes _UTF-8 String Literals_, and _Pattern matching with spans_ (specifically: `[ReadOnly]Span<char>`).

It would be interesting if both can be combined, allowing pattern matching against `[ReadOnly]Span<byte>`.

For example:

```cs
static HttpMethod ParseHttpMethod(ReadOnlySpan<byte> method)
{
    return method switch { 
        "GET"u8  => HttpMethod.Get, 
        "POST"u8 => HttpMethod.Post, 
...
        _ => HttpVerb.Unknown };
}
```

This is interesting for use-cases that involve parsing UTF-8.

cc @KathleenDollard

(Edited on 5/11/2022 by @stephentoub to include u8 suffix)

**Comment by tmds** [comment]
I've created an issue for this: https://github.com/dotnet/csharplang/issues/6037.

**Reply by jnm2** [reply]
This discussion thread rather than an issue is the appropriate place to have discussions. See the repo readme or the issue templates for guidance.

**Reply by RikkiGibson** [reply]
Yes, let's start by getting feedback from the community in this discussion thread. If the feedback is generally supportive here, we can look at opening an issue with a precisely outlined feature proposal, which the LDM can review.

**Comment by tmds** [comment]
It looks like this feature is mentioned in the [Test plan for Utf8StringLiterals feature](https://github.com/dotnet/roslyn/issues/58848/).

>  pattern matching ReadOnlySpan\<byte> against string constants

So, maybe this is already part of the plan for C# 11?

cc @AlekseyTs

**Comment by BreyerW** [comment]
I would definitely want this as i currently deal with json encoded in utf8 and being unable to use pattern matching would be both PITA and confusing when same thing works for utf-16 strings but for utf8 doesnt despite utf8 strings looking and even behaving like constant in pattern branches as long as it isnt interpolated (with exception being interpolated out of constants but support for this isnt necessary)

**Comment by tmds** [comment]
@AlekseyTs @KathleenDollard is this part of the plan for C# 11 already? If so, we can close the discussion.

**Comment by 333fred** [comment, marked as answer]
Discussed in LDM on May 11th, 2022. https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-05-11.md#pattern-matching-with-utf-8-string-literals

Conclusion: We will need a specification. I've opened https://github.com/dotnet/csharplang/issues/6161 for this, and put it in the any time milestone, which means a specification is open for community contribution.

**Comment by hez2010** [comment]
The priority of this feature definitely needs to be raised.
See how real-world first party application suffers from lacking pattern matching on UTF-8 strings: https://github.com/microsoft/garnet/blob/main/libs/server/Resp/RespCommand.cs

**Reply by 333fred** [reply]
See https://github.com/dotnet/csharplang/discussions/6036#discussioncomment-2816006.

**Reply by NinoFloris** [reply]
https://github.com/dotnet/csharplang/issues/6161 is locked for contributors at this point. You may want to teach the bot not to do that.

**Reply by jnm2** [reply]
That wasn't a bot! That was us following the repo's design process. Feature issues are locked because we don't want non-threaded discussions happening on them. Every feature issue points to a corresponding GitHub Discussions thread.
